Translation: "Using the State Hook" (#142)
reactjs/de.reactjs.org: content/docs/hooks-state.md

  render() {
    return (
      <div>
        <p>du hast mich {this.state.count} mal geklickt</p>
        <button onClick={() => this.setState({ count: this.state.count + 1 })}>
          Klick mich
        </button>
      </div>
    );
  }
}
```

Die State wird mit `{ count: 0 }` initialisiert. Über einen Klick auf den Button wird über den Aufruf `this.setState()` den Wert der State `state.count` inkrementiert. Nachfolgend werden wir auf der gesamten Seite Ausschnitte aus dieser Klasse verwenden.

>Hinweis
>
>Du fragst dich vielleicht, warum wir hier einen Zähler anstelle eines realistischeren Beispiels verwenden. Damit wollen wir uns vorerst auf die API konzentrieren, während wir noch unsere ersten Schritte mit den Hooks machen.

## Hooks und Funktionskomponenten {#hooks-and-function-components}

> Zur Erinnerung: In React sehen die Funktionskomponenten wie folgt aus:

Reviewer: ```suggestion
Zur Erinnerung: In React sehen die Funktionskomponenten so aus:
```

```js
const Example = (props) => {
  // Du kannst hier Hooks verwenden!
  return <div />;
}
```

> oder wie folgt:

Reviewer: ```suggestion
oder so:
```

}
```

> **Was ist ein Hook?** Ein Hook ist eine spezielle Funktion, mit der du dich in React-Funktionalitäten "einhaken" kannst. Betrachten wir uns beispielsweise die `useState`-Funktion an, sie ist ein Hook, mit dem wir States zu einer zustandslosen Komponente hinzufügen können, so dass wir die sogenannte Funktionskomponente erhalten. Wir werden später weitere Hooks kennenlernen.

Reviewer: ```suggestion
**Was ist ein Hook?** Ein Hook ist eine spezielle Funktion, mit der du dich in React-Funktionalitäten "einhaken" kannst. Schauen wir uns beispielsweise die `useState`-Funktion an, sie ist ein Hook, mit dem wir States zu einer zustandslosen Komponente hinzufügen können, so dass wir die sogenannte Funktionskomponente erhalten. Wir werden später weitere Hooks kennenlernen.
```

> **Wann sollte ich einen Hook verwenden?** Falls du bereits eine zustandslose Komponente verwendest und später feststellst, dass du die Komponente um States erweitern musst, dann musstest du früher deine Komponente in eine Klassenkomponente umschreiben, dies ist nicht mehr nötig, da du jetzt Hooks verwenden kannst. Diesen Schritt können wir jetzt in den Nachfolgenden Beispiel sehen!

Reviewer: ```suggestion
**Wann sollte ich einen Hook verwenden?** Falls du bereits eine zustandslose Komponente verwendest und später feststellst, dass du die Komponente um States erweitern musst, dann musstest du früher deine Komponente in eine Klassenkomponente umschreiben, dies ist nicht mehr nötig, da du jetzt Hooks verwenden kannst. Dies werden wir jetzt sofort tun!
```

  }
```

In einer Funktionskomponente existiert das Schlüsselwort `this` nicht, wodurch wir `this.state` weder zuweisen noch lesen können. Stattdessen rufen wir den `useState` Hook direkt in unserer Komponente auf:

```js{4,5}
import React, { useState } from 'react';

function Example() {

>   // Deklariere eine neue State Variable, die wir "count" nennen werden

Reviewer: ```suggestion
  // Deklariere eine neue State-Variable, die wir "count" nennen werden
```

  const [count, setCount] = useState(0);
```

> **Was bewirkt das Aufrufen der `useState` Funktion?** Es deklariert eine State-Variable. Die Variable heißt `count`, aber wir könnten sie beliebig nennen, wie z.B. `banane`. Dies ist eine Möglichkeit, einige Werte zwischen den Funktionsaufrufen zu "erhalten" — `useState` beschreibt eine neue Möglichkeit wie man exakt die selben Eigenschaften zu erhalten, wie sie `this.state` in einer Klasse zu finden sind. In der Regel "verschwinden" Variablen, wenn eine Funktion fertig ausgeführt ist, jedoch bleiben die State-Variabeln durch React erhalten.

Reviewer: ```suggestion
**Was bewirkt das Aufrufen der `useState`-Funktion?** Es deklariert eine "State-Variable". Die Variable heißt `count`, aber wir könnten sie beliebig nennen, wie z.B. `banane`. Dies ist eine Möglichkeit, einige Werte zwischen den Funktionsaufrufen zu "erhalten" — `useState` ist eine neue Möglichkeit wie man exakt die selben Eigenschaften zu benutezn, wie bei `this.state` in einer Klasse. In der Regel "verschwinden" Variablen, wenn eine Funktion fertig ausgeführt ist, jedoch bleiben die State-Variablen durch React erhalten.
```

> **Welche Argumente können wir an die `useState` Funktion übergeben?** Das einzige Argument, dass der `useState()` Hook verwendet ist der Anfangszustand. Anders als es bei Klassen der Fall ist, muss die State nicht unbedingt ein Objekt sein. Wir könnten z.B. eine Zahl oder eine Zeichenkette verwenden falls dies für uns ausreichend ist. In unserem Beispiel ist die verwendung einer Zahl ausreichend, da wir lediglich die Anzahl der Klicks auswerten möchten, demzufolge übergeben wir `0` als Anfangswert unserer State-Variable. (Falls wir zwei verschiedene Werte als State speichern wollen, dann rufen wir die `useState()` Funktion zweimal auf.)

Reviewer: ```suggestion
**Welche Argumente können wir an die `useState`-Funktion übergeben?** Das einzige Argument, dass der `useState()` Hook verwendet ist der initiale Zustand. Anders als es bei Klassen der Fall ist, muss der State nicht unbedingt ein Objekt sein. Wir könnten z.B. eine Zahl oder eine Zeichenkette verwenden falls dies für uns ausreichend ist. In unserem Beispiel ist die Verwendung einer Zahl ausreichend, da wir lediglich die Anzahl der Klicks auswerten möchten, demzufolge übergeben wir `0` als initialen Wert unserer State-Variable. (Falls wir zwei verschiedene Werte im State speichern wollen,  rufen wir `useState()` zweimal auf.)
```

> **Was gibt die `useState` Funktion zurück?** Es gibt ein Tupel zurück: die aktuelle State und eine Funktion die die aktuelle State aktualisieren kann. Aus diesem Grund schreiben wir `const [count, setCount] = useState()`. Dies ist vergleichbar mit `this.state.count` und `this.setState` die in Klassen verwendet werden, mit den einzigen Unterschied, dass man sie als Tupel erhält. Falls du mit der von uns verwendeten Syntax nicht vertraut bist, dann werden wir [am Ende dieser Seite](/docs/hooks-state.html#tip-what-do-square-brackets-mean) darauf zurückkommen.

Reviewer: ```suggestion
**Was gibt die `useState`-Funktion zurück?** Sie gibt ein Tupel zurück: der aktuelle State und eine Funktion die, den aktuellen State aktualisieren kann. Aus diesem Grund schreiben wir `const [count, setCount] = useState()`. Dies ist vergleichbar mit `this.state.count` und `this.setState` die in Klassen verwendet werden, mit dem einzigen Unterschied, dass man sie als Tupel erhält. Falls du mit der von uns verwendeten Syntax nicht vertraut bist, werden wir [am Ende dieser Seite](/docs/hooks-state.html#tip-what-do-square-brackets-mean) nochmal darauf zurückkommen.
```

Da wir nun wissen, was der `useState` Hook macht, sollte unser Beispiel mehr Sinn machen:

```js{4,5}
import React, { useState } from 'react';

function Example() {

>   // Deklariere eine neue State Variable, die wir "count" nennen werden

Reviewer: ```suggestion
  // Deklariere eine neue State-Variable, die wir "count" nennen werden
```

  const [count, setCount] = useState(0);
```

> Wir deklarieren eine State-Variable namens `count` und setzen sie auf `0`. React merkt sich den aktuellen Wert der Variable zwischen den re-rendering Phasen und stellt die aktuellste Version für unsere Funktion zur Verfügung. Falls wir den Wert der State-Variable `count` aktualisieren möchten, dann können wir die Methode `setCount` aufrufen.

Reviewer: ```suggestion
Wir deklarieren eine State-Variable namens `count` und setzen sie auf `0`. React merkt sich den aktuellen Wert der Variable zwischen dem neu Rendern und stellt uns diesen für unsere Funktion zur Verfügung. Falls wir den Wert der State-Variable `count` aktualisieren möchten, können wir die Methode `setCount` aufrufen.
```

>Hinweis
>

> >Mäglicherweise stellst du dir die nachfolgende Frage: Warum heißt die Hook `useState` an Stelle von `createState`?

Reviewer: ```suggestion
>Du fragst dich vielleicht: Warum heißt die Hook `useState` anstatt `createState`?
```
In this translation you use "die Hook", but in all other cases you use "der Hook". I think "die" sounds better. What do you think?

>

> >"Create" wäre nicht ganz zutreffend, da die State nur beim ersten Rendern der Komponente erstellt wird. Während der nachfolgenden Rendervorgänge stellt uns die `useState` Funktion die aktuelle State bereit. Andernfalls wäre es überhaupt kein "state"! Es hat auch seinen Grund, weshalb die Namen der Hooks *immer* mit `use` beginnen. Die Gründe dafür werden wir später in den [Regeln von Hooks](/docs/hooks-rules.html) erfahren.

Reviewer: ```suggestion
>"Create" wäre nicht ganz zutreffend, da der State nur beim ersten Rendern der Komponente erstellt wird. Während der nachfolgenden Rendervorgänge stellt uns die `useState`-Funktion den aktuellen State bereit. Andernfalls wäre es überhaupt kein "state"! Es hat auch seinen Grund, weshalb die Namen der Hooks *immer* mit `use` beginnen. Die Gründe dafür werden wir später in den [Regeln von Hooks](/docs/hooks-rules.html) erfahren.
```

> ## Lesen eines States {#reading-state}

Reviewer: ```suggestion
## Lesen des States {#reading-state}
```

> Um den aktuellen Zählerstand in einer Klasse anzeigen zulassen, dann lesen wir den Wert des Zählers über `this.state.count`:

Reviewer: ```suggestion
Wenn wir den aktuellen Zählerstand in einer Klasse anzeigen lassen wollen, lesen wir `this.state.count`:
```

```js

>   <p>du hast mich {this.state.count} mal geklickt</p>

Reviewer: ```suggestion
  <p>Du hast mich {this.state.count} mal geklickt</p>
```

```

In einer Funktion können wir `count` direkt verwenden:

```js

>   <p>du hast mich {count} mal geklickt</p>

Reviewer: ```suggestion
  <p>Du hast mich {count} mal geklickt</p>
```

```

> ## Aktualisierung eines States {#updating-state}

Reviewer: ```suggestion
## Aktualisieren des States {#updating-state}
```

> In einer Klasse müssen wir `this.setState()` aufrufen, um den `count` State zu aktualisieren:

Reviewer: ```suggestion
In einer Klasse müssen wir `this.setState()` aufrufen, um den `count`-State zu aktualisieren:
```

```js{1}
  <button onClick={() => this.setState({ count: this.state.count + 1 })}>
    Klick mich
  </button>
```

> In einer Funktion verfügen wir bereits über die `setCount` und `count` Variablen, so dass wir den Schlüsselwort `this` nicht benötigen:

Reviewer: ```suggestion
In einer Funktion haben wir bereits `setCount` und `count` als Variablen, so dass wir `this` nicht mehr benötigen:
```

```js{1}
  <button onClick={() => setCount(count + 1)}>
    Klick mich
  </button>
```

## Zusammenfassung {#recap}

> Wir möchten nun **das Gelernte Zeile für Zeile zusammenfassen** um den Stand unseres Wissens zu überprüfen.

Reviewer: ```suggestion
Wir möchten nun **das Gelernte Zeile für Zeile zusammenfassen** um unseren Wissensstand zu prüfen.
```

 5:
 6:    return (
 7:      <div>

>  8:        <p>du hast mich {count} mal geklickt</p>

Reviewer: ```suggestion
 8:        <p>Du hast mich {count} mal geklickt</p>
```

 9:        <button onClick={() => setCount(count + 1)}>
10:         Klick mich
11:        </button>
12:      </div>
13:    );
14:  }
```

> * **Zeile 1:** Wir importieren den `useState` Hook aus React, dadurch können wir eine State lokal in einer Funktionskomponente halten.

Reviewer: ```suggestion
* **Zeile 1:** Wir importieren den `useState` Hook aus React, dadurch können wir einen lokalen State in einer Funktionskomponente halten.
```

> * **Zeile 4:** Innerhalb der `Example`-Komponente deklarieren wir durch ein Aufruf des `useState` Hooks eine neue State-Variable, die ein Tupel zurückgibt, dessen Inhalte wir dann namen zuweisen. Indemfall nenen wir unsere Variable `count`, da sie die Anzahl der Button-klicks enthält. Am Anfang wird die Variable durch die Weitergabe von `0` als Argument an den `useState` Hook auf null initialisiert. Die zweite Variable die zurückgegeben wird ist eine funktion. Sie erlaubt uns den Wert von `count` zu aktualisieren, dementsprechend nennen wir sie `setCount`.

Reviewer: ```suggestion
* **Zeile 4:** Innerhalb der `Example`-Komponente deklarieren wir durch einen Aufruf des `useState`-Hooks eine neue State-Variable, die ein Tupel zurückgibt, dessen Inhalte wir dann Namen zuweisen. Wir nennen unsere Variable `count`, da sie die Anzahl der Button-Klicks enthält. Wir initialisieren Null, indem wir `0` als einziges Argument an `useState` geben. Die zweite Variable die zurückgegeben wird ist eine Funktion. Sie erlaubt uns den Wert von `count` zu aktualisieren, dementsprechend nennen wir sie `setCount`.
```

> * **Zeile 9:** Beim Klick auf den Button durch den User rufen wir die `setCount` Methode mit einem neuen Wert auf. React wird dann die `Example`-Komponente erneut rendern und übergibt der Koponente den neuen Wert für die `count` State durch.

Reviewer: ```suggestion
* **Zeile 9:** Beim Klick auf den Button durch den User rufen wir `setCount` mit einem neuen Wert auf. React wird dann die `Example`-Komponente erneut rendern und ihr den neuen Wert für den `count`-State übergeben.
```

> Auf den ersten Blick mag dies eine Menge sein. Am besten überstürtzt du nichts! Wenn du dich in der Fülle der Informationen verlierst, dann schau dir am besten den oben bereitgestellten Code nochmal genauer an und wiederhole die dargestellten Schritte von oben nach unten. Wir versprechen dir, dass falls du versuchst den Umgang mit den "Klassen" zu vergessen, dass du dann die Hooks mit neuen Augen sehen wirst und dadurch wird dir vieles aufeinmal Sinn machen.

Reviewer: ```suggestion
Auf den ersten Blick mag dies eine Menge sein. Überstürzte nichts! Wenn du dich in der Fülle der Informationen verlierst, dann schau dir am Besten den oben bereitgestellten Code nochmal genauer an und wiederhole die dargestellten Schritte von oben nach unten. Wir versprechen dir, dass wenn du versuchst zu "vergessen" wie der State in Klassen funktioniert und du den Code mit kühlem Kopf anschaust, es Sinn machen wird.
```

### Tipp: Welche Bedeutung haben die eckigen Klammern? {#tip-what-do-square-brackets-mean}

> Möglicherweise sind dir die eckigen Klammern aufgefallen, die wir bei der deklaration der State-Variabeln verwendet haben:

Reviewer: ```suggestion
Möglicherweise sind dir die eckigen Klammern aufgefallen, die wir bei der Deklaration der State-Variablen verwendet haben:
```

```js
  const [count, setCount] = useState(0);
```

Die Namen auf der linken Seite sind kein Bestandteil der React-API. Du kannst deine eigenen State-Variablen frei benennen:

```js
  const [fruit, setFruit] = useState('banane');
```

> Diese JavaScript-Syntax wird als ["Destrukturierende Array-Zuweisung"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring) bezeichnet. Dies bedeutet, dass wir zwei neue Variabeln `fruit` und `setFruit` erstellen, wobei `fruit` auf den ersten Wert gesetzt wird, der von den `useState`-Hook zurückgegeben wird, wogegen `setFruit` auf den zweiten Wert des Hooks gesetzt wird. Dies ist gleichwertig mit dem folgenden Code:

Reviewer: ```suggestion
Diese JavaScript-Syntax wird als ["Destrukturierende Array-Zuweisung"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring) bezeichnet. Dies bedeutet, dass wir zwei neue Variablen `fruit` und `setFruit` erstellen, wobei `fruit` auf den ersten Wert gesetzt wird, der von der `useState`-Hook zurückgegeben wird und `setFruit` auf den zweiten. Dies ist gleichwertig mit dem folgenden Code:
```

```js
  var fruitStateVariable = useState('banane'); // Gibt ein Tupel zurück
  var fruit = fruitStateVariable[0]; // Erstes Element im Tupel
  var setFruit = fruitStateVariable[1]; // Zweites Element im Tupel
```

> Bei der Deklaration einer State-Variable mithilfe von `useState` erhalten wir einen Tupel zurück — Indemfall ein Array mit zwei Elementen. Das erste Element ist der aktuelle Wert des States, Das zweite Element ist eine Funktion, mit der wir den Wert des States aktualisieren können. Die Verwendung von `[0]` und `[1]` um auf den Inhalt des Hooks zuzugreifen mag etwas verwirrend sein, dementsprechend verwenden wir die Array-Destrukturierung.

Reviewer: ```suggestion
Bei der Deklaration einer State-Variable mit Hilfe von `useState` erhalten wir einen Tupel zurück — ein Array mit zwei Werten. Der erste Wert ist der aktuelle des States, der zweite eine Funktion, mit der wir den Wert des States aktualisieren können. Die Verwendung von `[0]` und `[1]` um auf den Inhalt der Hook zuzugreifen mag etwas verwirrend sein, dementsprechend verwenden wir die Array-Destrukturierung.
```

>Hinweis
>
>Möglicherweise fragst du dich, woher React nun weiß, welche Komponente es `useState` zuordnen muss, da wir so etwas wie `this` nicht an React weitergeben. Wir werden die Antwort auf [diese Frage](/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components), sowie zahlreiche andere Fragen in der FAQ-Sektion beantworten.

> ### Tipp: Using Multiple State Variables {#tip-using-multiple-state-variables}

Reviewer: ```suggestion
### Tipp: Verwenden mehrerer State-Variablen {#tip-using-multiple-state-variables}
```

> Die Deklaration von State-Variablen als ein Tupel in der Form `[something, setSomething]` ist insofern praktisch, da es uns erlaubt die State-Variabeln *unterschiedliche* Namen zu geben, dies ist speziell Nützlich wenn man mehrere States verwenden möchte:

Reviewer: ```suggestion
Die Deklaration von State-Variablen als ein Tupel in der Form `[something, setSomething]` ist insofern praktisch, da es uns erlaubt den State-Variablen *unterschiedliche* Namen zu geben, dies ist nützlich wenn man mehrere States verwenden möchte:
```

```js
function ExampleWithManyStates() {

>   // Deklaration von mehreren States-Variablen!

Reviewer: ```suggestion
  // Deklaration von mehreren State-Variablen!
```

  const [age, setAge] = useState(42);
  const [fruit, setFruit] = useState('banane');
  const [todos, setTodos] = useState([{ text: 'Hooks erlernen' }]);
```

In der obigen Komponente stehen uns `age`, `fruit`, und `todos` als lokale Variablen zur Verfügung, die wir ebenfalls individuell aktualisieren können:

```js
  function handleOrangeClick() {
    // Vergleichbar mit this.setState({ fruit: 'orange' })
    setFruit('orange');
  }
```

> Du brauchst **nicht** zahlreichen State-Variablen zu verwenden. Die Variablen können problemlos Arrays oder Objekte halten, dementsprechend kannst du zusammenhängende Daten gruppieren. Beachte aber, dass im Gegensatz zu der in Klassen verwendeten Funktion `this.setState` die Aktualisierung durch die Hooks etwas anders funktioniert, da die State-Variable *ersetzt* wird, anstatt sie mit den bereits vorhandenen State zu verschmelzen.

Reviewer: ```suggestion
Du brauchst **nicht** zahlreichen State-Variablen zu verwenden. Die Variablen können problemlos Arrays oder Objekte halten, dementsprechend kannst du zusammenhängende Daten gruppieren. Beachte aber, dass im Gegensatz zu der in Klassen verwendeten Funktion `this.setState` die Aktualisierung durch die Hooks etwas anders funktioniert, da die State-Variable *ersetzt* wird, anstatt sie mit dem bereits vorhandenen State zu verschmelzen.
```

> Wir geben weitere Empfehlungen zur Aufspaltung unabhängiger State-Variablen die du [in den FAQ](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables) nachlesen kannst.

Reviewer: ```suggestion
Wir geben weitere Empfehlungen zur Aufteilung unabhängiger State-Variablen die du [in den FAQ](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables) nachlesen kannst.
```

## Nächste Schritte {#next-steps}

> Auf dieser Seite haben wir uns über einen der von React zur Verfügung gestellten Hooks, den sogenannten `useState` Hook informiert. Wir werden es gelegentlich auch als "State Hook" bezeichnen. Es ermöglicht uns die Funktionalität der Funktionskomponenten um States zu erweitern -- welches wir zum ersten Mal überhaupt getan haben!

Reviewer: ```suggestion
Auf dieser Seite haben wir uns über einen der von React zur Verfügung gestellten Hooks, den sogenannten `useState`-Hook. Wir werden es gelegentlich auch als "State Hook" bezeichnen. Es ermöglicht uns einen lokalen State zu einer React-Funktionskomponente hinzuzufügen -- welches wir zum ersten Mal überhaupt getan haben!
```

> Wir haben ebenfalls ein wenig mehr darüber erfahren, was Hooks überhaupt sind. Hooks sind Funktionen, mit denen man sich über Funktionskomponenten in React-Funktionalität "einhaken" kann. Ihr name beginnt immer mit `use`. Es gibt zahlreiche weitere Hooks, die wir noch nicht kennengelernt haben.

Reviewer: ```suggestion
Wir haben ebenfalls ein wenig mehr darüber erfahren, was Hooks überhaupt sind. Hooks sind Funktionen, mit denen man sich über Funktionskomponenten in die React-Funktionalität "einhaken" kann. Ihr Name beginnt immer mit `use`. Es gibt zahlreiche weitere Hooks, die wir noch nicht kennengelernt haben.
```

> **Lasst uns nun fortfahren in dem wir uns den [nächsten Hook widmen: `useEffect`.](/docs/hooks-effect.html)** Es ermöglicht dir die Durchführung von Effekten innerhalb der Komponenten, dies ist vergleichbar mit den Lifecycle-Methoden, die man in Klassen verwendet.

Reviewer: ```suggestion
**Fahren wir nun fort mit der [nächsten Hook: `useEffect`.](/docs/hooks-effect.html)** Diese ermöglicht dir die Durchführung von Seiten-Effekten innerhalb von Komponenten und ähnelt den Lifecycle-Methoden in Klassen.
```
